### PR TITLE
Set transaction payer intrinsic implementation

### PR DESCRIPTION
**Pull request by venu-block1**
## Change Description

Set transaction payer intrinsic implementation for the Transaction Sponsorship feature.

## Change Type
**Select *ONE*:**
- [ ] Documentation

- [ ] Stability bug fix

- [x] Other

- [ ] Other - special case

## Testing Changes
**Select *ANY* that apply:**
- [ ] New Tests

- [ ] Existing Tests

- [ ] Test Framework

- [ ] CI System

- [ ] Other

## Consensus Changes
- [ ] Consensus Changes

## API Changes
- [ ] API Changes

## Documentation Additions
- [x] Documentation Additions

Please add the new intrinsic to contract developers documentation.

### CODE REVIEW

File: libraries/chain/genesis_intrinsics.cpp [hunk 1]
```diff
@@ -55,6 +55,7 @@ const std::vector<const char*> genesis_intrinsics = {
    "set_blockchain_parameters_packed",
    "is_privileged",
    "set_privileged",
+   "set_transaction_resource_payer",
```
Comment (heifner) [inline libraries/chain/genesis_intrinsics.cpp:58]: This is not a genesis intrinsic. Model off of something like `set_proposed_producers_ex`

File: libraries/chain/include/eosio/chain/transaction.hpp [hunk 2]
```diff
@@ -61,6 +61,8 @@ namespace eosio { namespace chain {
       void set_reference_block( const block_id_type& reference_block );
       bool verify_reference_block( const block_id_type& reference_block )const;
       void validate()const;
+      void set_max_net( const uint64_t max_net ) { max_net_usage_words = max_net; }
+      void set_max_cpu( const uint64_t max_cpu ) { max_cpu_usage_ms = max_cpu; }
```
Comment (heifner) [inline libraries/chain/include/eosio/chain/transaction.hpp:65]: Should make it clear if max_cpu is milliseconds or microseconds.

File: libraries/chain/include/eosio/chain/webassembly/interface.hpp [hunk 3]
```diff
@@ -306,6 +306,17 @@ namespace webassembly {
...
          void set_privileged(account_name account, bool is_priv);
 
+         /**
+          * Set the resource payer for the transaction.
+          *
+          * @ingroup privileged
+          *
+          * @param payer - name of the account who is paying for the transaction.
+          * @param max_net - max amount of net resource to be paid for the transaction.
+          * @param max_cpu - max amount of cpu resource to be paid for the transaction.
```
Comment (heifner) [inline libraries/chain/include/eosio/chain/webassembly/interface.hpp:316]: Make it clear if this is microseconds or milliseconds. I believe the requirements are to support microseconds unlike the current transaction max_cpu_usage_ms which is milliseconds.

File: libraries/chain/webassembly/privileged.cpp [hunk 4]
```diff
@@ -287,4 +287,20 @@ namespace eosio { namespace chain { namespace webassembly {
          ma.set_privileged( is_priv );
       });
    }
+
+   bool interface::set_transaction_resource_payer( const name payer, const uint64_t max_net, const uint64_t max_cpu ) {
+      //  check if the payer has signed 
+      if ( has_auth ( payer ) )
+      {
+         transaction& trx = const_cast<transaction&>( context.trx_context.packed_trx.get_transaction() );
```
Comment (heifner) [inline libraries/chain/webassembly/privileged.cpp:295]: `transaction` is currently thread safe and needs to remain thread safe. Do not modify transaction data.

**Review by heifner — CHANGES_REQUESTED** [review]

File: libraries/chain/genesis_intrinsics.cpp (hunk 1 above)
Comment (venu-block1) [reply libraries/chain/genesis_intrinsics.cpp]: Removed from here.

File: libraries/chain/include/eosio/chain/transaction.hpp (hunk 2 above)
Comment (venu-block1) [reply libraries/chain/include/eosio/chain/transaction.hpp]: These set functions are moved to transaction context, and the units are made explicit.

File: libraries/chain/include/eosio/chain/webassembly/interface.hpp (hunk 3 above)
Comment (venu-block1) [reply libraries/chain/include/eosio/chain/webassembly/interface.hpp]: Done

File: libraries/chain/webassembly/privileged.cpp (hunk 4 above)
Comment (venu-block1) [reply libraries/chain/webassembly/privileged.cpp]: Ok, not touching transaction data. Made changes.

**Review by ndcgundlach — APPROVED** [review]

**Pull request merged**